syntax = "proto3";

package infrastructure.rpc.order.v1.model.v1;

option go_package = "github.com/shortlink-org/shop/oms/internal/infrastructure/rpc/order/v1/model/v1";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";
import "domain/order/v1/common/v1/common.proto";

// Define the Order message
message OrderState {
  // Unique identifier for the order
  string id = 1;
  // Identifier for the customer who placed the order
  string customer_id = 2;
  // List of items in the order
  repeated OrderItem items = 3;
  // Current status of the order
  domain.order.common.v1.OrderStatus status = 4;
  // Timestamp for when the order was created
  google.protobuf.Timestamp created_at = 5;
  // Timestamp for when the order was last updated
  google.protobuf.Timestamp updated_at = 6;
  // Delivery information (optional, nil = self-pickup)
  domain.order.common.v1.DeliveryInfo delivery_info = 7;
}

// Define the OrderItem message
message OrderItem {
  // Unique identifier for the item
  string id = 1;
  // Quantity of the item ordered
  int32 quantity = 2;
  // Price of a single item
  double price = 3;
}

// Request message for creating an order
message CreateRequest {
  // Order details
  OrderState order = 1;
  // Delivery information (optional, nil = self-pickup)
  domain.order.common.v1.DeliveryInfo delivery_info = 2;
}

// Request message for getting an order by ID
message GetRequest {
  // ID of the order to retrieve
  string id = 1;
}

// Response message for getting an order by ID
message GetResponse {
  // Retrieved order details
  OrderState order = 1;
}

// Request message for updating an order
message UpdateRequest {
  // Updated order details
  OrderState order = 1;
  // Fields to update
  google.protobuf.FieldMask update_mask = 2;
}

// Request message for deleting an order by ID
message CancelRequest {
  // ID of the order to delete
  string id = 1;
}

// Request message for updating delivery info
message UpdateDeliveryInfoRequest {
  // ID of the order to update
  string order_id = 1;
  // New delivery information
  domain.order.common.v1.DeliveryInfo delivery_info = 2;
}

// Request message for checkout (creating order from cart)
message CheckoutRequest {
  // ID of the customer whose cart should be converted to order
  string customer_id = 1;
  // Delivery information (optional, nil = self-pickup)
  domain.order.common.v1.DeliveryInfo delivery_info = 2;
}

// Response message for checkout
message CheckoutResponse {
  // ID of the created order
  string order_id = 1;
}

// Pagination info for list requests
message Pagination {
  // Current page number (1-indexed)
  int32 page = 1;
  // Number of items per page
  int32 page_size = 2;
}

// Pagination info for list responses
message PaginationResponse {
  // Current page number
  int32 current_page = 1;
  // Number of items per page
  int32 page_size = 2;
  // Total number of pages
  int32 total_pages = 3;
}

// Request message for listing orders
message ListRequest {
  // Optional filter by customer ID
  string customer_id = 1;
  // Optional filter by order status
  repeated domain.order.common.v1.OrderStatus status_filter = 2;
  // Pagination
  Pagination pagination = 3;
}

// Response message for listing orders
message ListResponse {
  // List of orders
  repeated OrderState orders = 1;
  // Total count of orders matching the filter
  int32 total_count = 2;
  // Pagination info
  PaginationResponse pagination = 3;
}
